newton/utils/render.py

```diff
@@ -481,7 +483,302 @@ class SimRendererUsd(CreateSimRenderer(renderer=UsdRenderer)):
+            scaling (float, optional): Scaling factor for the rendered objects. Defaults to 1.0.
+            fps (int, optional): Frames per second for the animation. Defaults to 60.
+            up_axis (newton.AxisType, optional): Up axis for the scene. If None, uses model's up axis.
+            show_joints (bool, optional): Whether to show joint visualizations.  Defaults to False.
+            path_body_map (dict, optional): A dictionary mapping prim paths to body IDs.
+            path_body_relative_transform (dict, optional): A dictionary mapping prim paths to relative transformations.
+            builder_results (dict, optional): A dictionary containing builder results.
+            **render_kwargs: Additional arguments passed to the underlying UsdRenderer.
+        """
+        if source_stage:
+            if path_body_map is None:
+                raise ValueError("path_body_map must be set if you are providing a source_stage")
+            if path_body_relative_transform is None:
+                raise ValueError("path_body_relative_transform must be set if you are providing a source_stage")
+            if builder_results is None:
+                raise ValueError("builder_results must be set if you are providing a source_stage")
+
+        self.source_stage = source_stage
+
+        if not self.source_stage:
+            super().__init__(
+                model=model,
+                path=stage,
+                scaling=scaling,
+                fps=fps,
+                up_axis=up_axis,
+                show_joints=show_joints,
+                **render_kwargs,
+            )
+        else:
+            self.stage = self._create_output_stage(self.source_stage, stage)
+            self.fps = fps
+            self.scaling = scaling
+            self.up_axis = up_axis
+            self.path_body_map = path_body_map
+            self.path_body_relative_transform = path_body_relative_transform
+            self.builder_results = builder_results
+            self._prepare_output_stage()
+            self._precompute_parents_xform_inverses()
+
+    def render_update_stage(self, state: newton.State):
+        if not self.source_stage:
+            raise ValueError("source_stage must be set before calling render_update_stage")
+
+        self._update_usd_stage(state)
+
+    def _update_usd_stage(self, state: newton.State):
+        """
+        Render transforms of USD prims as time-sampled animation in USD.
+
+        Args:
+            state (newton.State): The simulation state to render.
+            sim_time (float): The current simulation time.
+        """
+        from pxr import Sdf
+
+        body_q = state.body_q.numpy()
+        with Sdf.ChangeBlock():
+            for prim_path, body_id in self.path_body_map.items():
+                full_xform = body_q[body_id]
+                # TODO: do this once in __init__
+                # TODO: sanity check this with Eric Heiden
```

Comment:
Are these TODO's still relevant?

Reply:
I'm leaving out 1 `TODO` for a more efficient implementation to be addressed at a later stage.